src/category_theory/groupoid/subgroupoid.lean

```diff
@@ -0,0 +1,356 @@
+
+
+## Main definitions
+
+Given a type `C` with associated `groupoid C` instance.
+
+* `subgroupoid C` is the type of subgroupoids of `C`
+* `subgroupoid.is_normal` is the property that the subgroupoid is stable under conjugation
+  by arbitrary arrows, _and_ that all identity arrows are contained in the subgroupoid.
+* `subgroupoid.comap` is the "preimage" map of subgroupoids along a functor.
+* `subgroupoid.map` is the "image" map of subgroupoids along a functor _injective on objects_.
+* `subgroupoid.vertex_subgroup` is the subgroup of the `vertex group` at a given vertex `v`,
+  assuming `v` is contained in the `subgroupoid` (meaning, by definition, that the arrow `𝟙 v`
+  is contained in the subgroupoid).
+
+## Implementation details
+
+The structure of this file is copied from/inspired by `group_theory.subgroup.basic`
+and `combinatorics.simple_graph.subgraph`.
+
+## TODO
+
+* Equivalent inductive characterization of generated (normal) subgroupoids.
+* Characterization of normal subgroupoids as kernels.
+
+## Tags
+
+subgroupoid
+
+-/
+
+open set classical function
+
+namespace category_theory
+
+open groupoid
+
+local attribute [protected] category_theory.inv
+
+universes u v
+
+variables {C : Type u} [groupoid C]
+
+/--
+A sugroupoid of `C` consists of a choice of arrows for each pair of vertices, closed
+under composition and inverses
+-/
+@[ext] structure subgroupoid (C : Type u) [groupoid C] :=
+(arrows : ∀ (c d : C), set (c ⟶ d))
+(inv : ∀ {c d} {p : c ⟶ d} (hp : p ∈ arrows c d),
+          inv p ∈ arrows d c)
+(mul : ∀ {c d e} {p} (hp : p ∈ arrows c d) {q} (hq : q ∈ arrows d e),
+          p ≫ q ∈ arrows c e)
+
+attribute [protected] subgroupoid.inv
+attribute [protected] subgroupoid.mul
+
+namespace subgroupoid
+
+variable (S : subgroupoid C)
+
+/-- The vertices of `C` on which `S` has non-trivial isotropy -/
+def objs : set C := {c : C | (S.arrows c c).nonempty}
+
+lemma id_mem_of_nonempty_isotropy (c : C) :
+  c ∈ objs S → 𝟙 c ∈ S.arrows c c :=
+begin
+  rintro ⟨γ,hγ⟩,
+  convert S.mul hγ (S.inv hγ),
+  simp only [inv_eq_inv, is_iso.hom_inv_id],
+end
+
+/-- A subgroupoid seen as a quiver on vertex set `C` -/
+def as_wide_quiver : quiver C := ⟨λ c d, subtype $ S.arrows c d⟩
+
+/-- The coercion of a subgroupoid as a groupoid -/
+instance coe : groupoid S.objs :=
+{ hom := λ a b, S.arrows a.val b.val,
+  id := λ a, ⟨𝟙 a.val, id_mem_of_nonempty_isotropy S a.val a.prop⟩,
+  comp := λ a b c p q, ⟨p.val ≫ q.val, S.mul p.prop q.prop⟩,
+  id_comp' := λ a b ⟨p,hp⟩, by simp only [category.id_comp],
+  comp_id' := λ a b ⟨p,hp⟩, by simp only [category.comp_id],
+  assoc' := λ a b c d ⟨p,hp⟩ ⟨q,hq⟩ ⟨r,hr⟩, by simp only [category.assoc],
+  inv := λ a b p, ⟨inv p.val, S.inv p.prop⟩,
+  inv_comp' := λ a b ⟨p,hp⟩, by simp only [inv_comp],
+  comp_inv' := λ a b ⟨p,hp⟩, by simp only [comp_inv] }
+
+/-- There is an embedding of the coerced subgroupoid to its parent-/
+def hom : S.objs ⥤ C :=
```

Comment:
This is the analogue of [subgroup.subtype](https://leanprover-community.github.io/mathlib_docs/group_theory/subgroup/basic.html#subgroup.subtype) but I think the current name is also okay.